librz/analysis/arch/arm/arm_il32.c

```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
 	return JMP(ADD(U32(PC(insn->address, is_thumb)), SHIFTL0(UNSIGNED(32, off), UN(5, 1))));
 }
 
+static RzILOpEffect *write_reg_lane(arm_reg reg, ut32 lane, ut32 vec_size, RzILOpBitVector *v) {
+	ut32 reg_size = reg_bits(reg);
+	RzILOpBitVector *sft_val = SHIFTL0(UNSIGNED(reg_size, v), UN(8, vec_size * lane));
+	return write_reg(reg, sft_val);
+}
+
+/**
+ * For Extend Instruction Set
+ * TODO: Split to a seperate file and include before arm lifter
+ * VFP and NEON
+ */
+
 /**
  * Capstone: ARM_INS_VMOV
  * ARM: vmov
  */
 static RzILOpEffect *vmov(cs_insn *insn, bool is_thumb) {
-	if (OPCOUNT() != 2 || !ISREG(0) || !ISREG(1)) {
+	if (OPCOUNT() < 2 || !ISREG(0) || (!ISIMM(1) && !ISREG(1))) {
 		// for now, only support vmov rt, rn
 		return NULL;
 	}
+
+	// vmov for immediate must be unconditional
+	// vmov.<dt> <Qd>, or vmov.<dt> <Dd>
+	// vmov.F32 <Sd>, or vmov.F64 <Qd>
+	if (ISIMM(1)) {
+		// possible : I8, I16, I32, I64, F32 for Q/D register
+		// possible : F32 for S register
+		// possible : F64 for D register
+		ut32 imm_width = DT_WIDTH(insn);
+		if (!imm_width) {
+			return NULL;
+		}
+		ut32 reg_width = REG_WIDTH(0);
+		if (reg_width < imm_width) {
+			return NULL;
+		}
+
+		ut32 imm = get_imm(insn, 1, NULL);
+		RzILOpBitVector *imm_bv = repeated_imm(imm_width, reg_width, imm);
+		if (!imm_bv) {
+			return NULL;
+		}
+
+		// vmvn Dd, imm
+		if (insn->id == ARM_INS_VMVN) {
+			imm_bv = LOGNOT(imm_bv);
+		}
+
+		return write_reg(REGID(0), imm_bv);
+	}
+
+	// 2 core registers and 1 double-word register
+	if (OPCOUNT() == 3) {
+		if (!is_core_reg(REGID(0))) {
+			// vmov <Dm> <Rt1> <Rt2>, Dm[low] = Rt1, Dm[high] = Rt2
+			RzILOpBitVector *rt1_val = REG(1);
+			RzILOpBitVector *rt2_val = REG(2);
+			if (!rt1_val || !rt2_val) {
+				return NULL;
+			}
+			return write_reg(REGID(0), APPEND(rt2_val, rt1_val));
+		}
+		// vmov <Rt1> <Rt2> <Dm>, Rt1 = Dm[low], Rt2 = Dm[high]
+		RzILOpBitVector *reg_val = REG(2);
+		if (!reg_val) {
+			return NULL;
+		}
+		RzILOpBitVector *rt1_val = UNSIGNED(32, reg_val);
+		RzILOpBitVector *rt2_val = UNSIGNED(32, SHIFTR0(UN(8, 32), reg_val));
+		return SEQ2(write_reg(REGID(0), rt1_val),
+			write_reg(REGID(1), rt2_val));
+	}
+
+	// 2 core registers and 2 single-word registers
+	// vmov <Sm1> <Sm2> <Rt1> <Rt2>
+	// vmov <Rt1> <Rt2> <Sm1> <Sm2>
+	if (OPCOUNT() == 4) {
+		RzILOpBitVector *rt1_val = REG(2);
+		RzILOpBitVector *rt2_val = REG(3);
+		if (!rt1_val || !rt2_val) {
+			return NULL;
+		}
+		return SEQ2(write_reg(REGID(0), rt1_val),
+			write_reg(REGID(1), rt2_val));
+	}
+
+	// core register to scalar
+	if (NEON_LANE(0) != -1 && !is_core_reg(REGID(0)) && is_core_reg(REGID(1))) {
+		// vmov.<vec_size> <Dd>[lane], <Rt>
+		// <Dd>[lane] = <Rt>{vecsize - 1 : 0}
+		if (!VVEC_SIZE(insn) || NEON_LANE(0) == -1) {
+			return NULL;
```

Comment:
`rz_warn_if_reached()`?

Reply:
Same below. Basically whenever `NULL` is returned and this means something unexpected happened.

Reply:
And this one